steps/terraform.yaml

```diff
@@ -0,0 +1,152 @@
+parameters:
+  - name: adolibrary
+    default: cnp-azuredevops-libraries
+
+  - name: initCommandOptions
+    default: ''
+
+  - name: planCommandOptions
+    default: ''
+
+  - name: applyCommandOptions
+    default: ''
+
+  - name: destroyCommandOptions
+    default: ''
+
+  - name: environment
+    default: sandbox
+
+  - name: location
+    default: 'UK South'
+    values:
+      - 'UK South'
+      - 'UK West'
+
+  - name: overrideAction
+    default: plan
+    values:
+      - plan
+      - apply
+      - destroy
+
+  - name: project
+    default: ''
+    values:
+      - cft
+      - ss
+
+  - name: product
+    default: ''
+
+  - name: stack
+    default: ''
+
+  - name: serviceConnection
+    default: ''
+
+  - name: tfversion
+    default: ''
+
+  - name: tfInitSub
+    default: ''
+
+steps:
+  - checkout: self
+  - checkout: ${{ parameters.adoLibrary }}
+  - bash: echo ${{ parameters.environment }}
+
+  - template: ./set-build-repo-suffix-env-var.yaml
+  - template: ./keyvault-read.yaml
+    parameters:
+      serviceConnection: ${{ parameters.serviceConnection }}
+      environment: ${{ parameters.environment }}
+
+  - task: TerraformInstaller@0
+    displayName: Terraform install
+    inputs:
+      terraformVersion: ${{ parameters.tfversion }}
+
+  - bash: |
+      echo "repo short name - $(buildRepoSuffix)"
+      echo "working dir - $(System.DefaultWorkingDirectory)/$(buildRepoSuffix)/components/${{ parameters.stack }}"
+      tree "$(System.DefaultWorkingDirectory)/$(buildRepoSuffix)/components/${{ parameters.stack }}"
+
+  - task: TerraformCLI@0
+    displayName: Terraform init ${{ parameters.stack }}
+    inputs:
+      command: init
+      terraformVersion: ${{ parameters.tfversion }}
+      workingDirectory: '$(System.DefaultWorkingDirectory)/$(buildRepoSuffix)/components/${{ parameters.stack }}'
+      backendType: azurerm
+      ensureBackend: false
+      backendServiceArm: ${{ parameters.serviceConnection }}
+      ${{ if eq(parameters['environment'], 'ptl') }}:
+        backendAzureRmResourceGroupName: "azure-control-prod-rg"
+      ${{ if ne(parameters['environment'], 'ptl') }}:
+        backendAzureRmResourceGroupName: 'azure-control-${{ parameters.environment }}-rg'
+      backendAzureRmResourceGroupLocation: ${{ parameters.location }}
+      backendAzureRmStorageAccountName: $(controlStorageAccount)
+      backendAzureRmStorageAccountSku: Standard_LRS
+      backendAzureRmContainerName: subscription-tfstate
+      backendAzureRmKey: '${{ parameters.location }}/${{ parameters.project }}/${{ parameters.product }}/$(buildRepoSuffix)/${{ parameters.environment }}/${{ parameters.stack }}/terraform.tfstate'
+      commandOptions: '-backend-config=subscription_id=${{ parameters.tfInitSub }} ${{ parameters.initCommandOptions }}'
+
+  - task: TerraformInstaller@0
```

Comment:
Should we have this validation step in the pre-check stage? 
So all the validations can be done at the pre-check stage and once that stage passes all we have is to plan and apply.